File: api_sw/app/Plugin/Sms/AliyunSms.php
<?php

declare(strict_types=1);

namespace App\Plugin\Sms;

use AlibabaCloud\SDK\Dysmsapi\V20170525\Dysmsapi;
use \Exception;
use AlibabaCloud\Tea\Exception\TeaError;
use AlibabaCloud\Tea\Utils\Utils;
use Darabonba\OpenApi\Models\Config;
use AlibabaCloud\SDK\Dysmsapi\V20170525\Models\SendSmsRequest;
use AlibabaCloud\Tea\Utils\Utils\RuntimeOptions;

class AliyunSms extends AbstractSms
{
    protected $config = [];

    /**
     * 发送短信
     *
     * @param string $phone
     * @param string $code
     * @return void
     */
    public function send(string $phone, string $code)
    {
        return $this->sendSms([$phone], json_encode(['code' => $code], JSON_UNESCAPED_UNICODE));
    }

    /**
     * 发送短信(批量)
     *
     * @param array $phoneArr
     * @param string $templateParam
     * @return void
     */
    public function sendSms(array $phoneArr, string $templateParam)
    {
        $client = $this->createClient();
        $sendSmsRequest = new SendSmsRequest([
            'phoneNumbers' => 'your_value',
            'signName' => $this->config['signName'],
            'templateCode' => $this->config['templateCode'],
            'templateParam' => $templateParam,
            // 'templateParam' => '{"code": "1234"}',
        ]);
        try {
            $result = $client->sendSmsWithOptions($sendSmsRequest, new RuntimeOptions([]));
            $result = $result->toMap();
            if (!(isset($result['body']['Code']) && $result['body']['Code'] == 'OK')) {
                throwFailJson(79999999, '阿里云SMS错误：' . $result['body']['Message']);
            }
        } catch (Exception $error) {
            if (!($error instanceof TeaError)) {
                $error = new TeaError([], $error->getMessage(), $error->getCode(), $error);
            }
            $errMsg = Utils::assertAsString($error->message);
            throwFailJson(79999999, '阿里云SMS错误：' . $errMsg);
        }
    }

    /**
     * 使用AK&SK初始化账号Client
     * @param string $accessKeyId
     * @param string $accessKeySecret
     * @return Dysmsapi Client
     */
    public function createClient()
    {
        $config = new Config([
            'accessKeyId' => $this->config['accessKeyId'],
            'accessKeySecret' => $this->config['accessKeySecret']
        ]);
        $config->endpoint = 'dysmsapi.aliyuncs.com';
        return new Dysmsapi($config);
    }
}

File: api/internal/utils/sms/aliyun_sms.go
package sms

import (
	"context"
	"errors"
	"strings"

	openapi "github.com/alibabacloud-go/darabonba-openapi/v2/client"
	dysmsapi20170525 "github.com/alibabacloud-go/dysmsapi-20170525/v3/client"
	util "github.com/alibabacloud-go/tea-utils/v2/service"
	"github.com/alibabacloud-go/tea/tea"
	"github.com/gogf/gf/v2/util/gconv"
)

type AliyunSms struct {
	Ctx             context.Context
	AccessKeyId     string `json:"aliyunSmsAccessKeyId"`
	AccessKeySecret string `json:"aliyunSmsAccessKeySecret"`
	Endpoint        string `json:"aliyunSmsEndpoint"`
	SignName        string `json:"aliyunSmsSignName"`
	TemplateCode    string `json:"aliyunSmsTemplateCode"`
}

func NewAliyunSms(ctx context.Context, config map[string]interface{}) *AliyunSms {
	aliyunSmsObj := AliyunSms{
		Ctx: ctx,
	}
	gconv.Struct(config, &aliyunSmsObj)
	return &aliyunSmsObj
}

func (smsThis *AliyunSms) Send(phone string, code string) (err error) {
	err = smsThis.SendSms([]string{phone}, `{"code": "`+code+`"}`)
	return
}

func (smsThis *AliyunSms) SendSms(phoneArr []string, templateParam string) (err error) {
	client, err := smsThis.CreateClient()
	if err != nil {
		return
	}

	sendSmsRequest := &dysmsapi20170525.SendSmsRequest{
		PhoneNumbers:  tea.String(strings.Join(phoneArr, `,`)),
		SignName:      tea.String(smsThis.SignName),
		TemplateCode:  tea.String(smsThis.TemplateCode),
		TemplateParam: tea.String(templateParam),
		// TemplateParam: tea.String(`{"code": "1234"}`),
	}

	tryErr := func() (_e error) {
		defer func() {
			r := tea.Recover(recover())
			if r != nil {
				_e = r
			}
		}()
		result, err := client.SendSmsWithOptions(sendSmsRequest, &util.RuntimeOptions{})
		if err != nil {
			return err
		}
		if *result.Body.Code != `OK` {
			err = errors.New(*result.Body.Message)
			return err
		}
		return nil
	}()

	if tryErr != nil {
		var errSDK = &tea.SDKError{}
		_t, ok := tryErr.(*tea.SDKError)
		if ok {
			errSDK = _t
		} else {
			errSDK.Message = tea.String(tryErr.Error())
		}
		errMsg, errSms := util.AssertAsString(errSDK.Message)
		if errSms != nil {
			err = errSms
			return
		}
		err = errors.New(*errMsg)
	}
	return
}

func (smsThis *AliyunSms) CreateClient() (client *dysmsapi20170525.Client, err error) {
	config := &openapi.Config{
		AccessKeyId:     tea.String(smsThis.AccessKeyId),
		AccessKeySecret: tea.String(smsThis.AccessKeySecret),
		Endpoint:        tea.String(smsThis.Endpoint),
	}
	client, err = dysmsapi20170525.NewClient(config)
	return
}
